### PR TITLE
INFO ONLY: initial drush9 conversion

### PR DESCRIPTION
**Pull request by rivimey**
Ken, thought I'd show you where I'm at with the drush9 conversion of the drush commands.
Most things work except for the delete domain and some forms of verify. Work continuing...

I've changed some of the other interfaces where that seemed reasonable to me. Please let me know if these are not viable. These are in : DomainStorage. DomainValidator

### CODE REVIEW

File: domain/src/DomainStorage.php [hunk 1]
```diff
@@ -82,6 +82,8 @@ public function loadDefaultId() {
 
   /**
    * {@inheritdoc}
+   *
+   * @return \Drupal\domain\DomainInterface|NULL
```
Comment (zerolab) [inline domain/src/DomainStorage.php]: these returns should go on the interface

Comment (rivimey) [reply domain/src/DomainStorage.php]: Thx for looking!
I've just reviewed this, and these are serving a useful purpose: the interface does have these declarations, but the class {@inheritdoc} is not inheriting the interface but the base class doc, which has EntityInterface.

The net result is that within code manipulating class object refs (e.g. within domain module) these functions return EntityInterface objects, not DomainInterface objects.

I've pushed an update which works for me: let me know if there's a better way.

File: domain/src/DomainStorage.php [hunk 2]
```diff
@@ -98,7 +98,7 @@ public function loadDefaultDomain() {
    * @return \Drupal\domain\DomainInterface[]
    */
   public function loadMultiple(array $ids = NULL) {
-    $domains = $this->loadMultiple($ids);
+    $domains = parent::loadMultiple($ids);
```
Comment (agentrickard) [inline domain/src/DomainStorage.php]: If we're calling parent here, do we even need this method override? 

This may be a legacy method from before EntityStorage was standardized.

Comment (rivimey) [reply domain/src/DomainStorage.php]: This and some others are my fault: I wanted the functions in DomainStorage class to be declared as returning items of type DomainInterface, so the type checking in phpcs correctly reports on functions and their args.

I am very happy to remove the additional code if there is a better way. Might that way be using @method tags: http://docs.phpdoc.org/references/phpdoc/tags/method.html / https://www.drupal.org/project/coding_standards/issues/2920333 ?

If so the whole loadMultiple function and docblock referenced here could, I think, be reduced to:

    /**
     * @method \Drupal\domain\DomainInterface[] loadMultiple(array $ids = NULL)
     */

Comment (agentrickard) [reply domain/src/DomainStorage.php]: The use of parent here is a code smell.

I wouldn't rewrite that, because it confuses developers used to the core API. Returning generic entities are fine. I think you can still type hint on the array elements returned by EntityStorageBase::loadMultiple() if you wanted to.

The `sort` method called by LoadMultipleSorted() does that.

```
  /**
   * {@inheritdoc}
   */
  public function sort(DomainInterface $a, DomainInterface $b) {
    return $a->getWeight() > $b->getWeight();
  }
```

Comment (rivimey) [reply domain/src/DomainStorage.php]: I agree about the smell, but the case of sort() here is for parameters, not function results.

Just to be sure, did you realize that I added the whole loadMultiple() function around line 100, not just changed line 101?

I don't really understand "confuses developers used to the core API", given that these functions are being called on domain objects and have the same function signature and results as for the entity api: could you elaborate?

I wonder if it would be useful to meet on drupal slack tomorrow to clarify this and any other issues?

Comment (agentrickard) [reply domain/src/DomainStorage.php]: Sure. I can be in Slack. I'm around today, too.

I only noticed you added this method after the initial reply. 

The sort method definition type hints the array elements returned by loadMultiple(), knowing that the base class will return DomainInterface objects as an array.

```
 public function sort(DomainInterface $a, DomainInterface $b)
```

The confusion is that our class has a method that overwrites the base class but merely invokes it as `parent`, which makes me scratch my head as a developer. Why override the base method if it isn't changed?

File: domain/src/DomainStorage.php [hunk 3]
```diff
@@ -94,6 +95,16 @@ public function loadDefaultDomain() {
 
   /**
    * {@inheritdoc}
+   * @return \Drupal\domain\DomainInterface[]
+   */
+  public function loadMultiple(array $ids = NULL) {
```
Comment (zerolab) [inline domain/src/DomainStorage.php]: I really think these do not belong here. 
`NodeStorage` does not do this, nor other things in core.
As @agentrickard said, you are merely deferring to the parent and that is an antipattern as far as I kow

**Comment by rivimey** [comment]
No problem, I've pulled all the changes to DomainStorage*

File: domain/src/Commands/DomainCommands.php [hunk 4]
```diff
@@ -0,0 +1,1108 @@
...
+
+  /**
+   * Get a domain storage object or throw an exception.
+   *
+   * @return \Drupal\domain\DomainStorageInterface
+   *
+   * @throws \Drupal\domain\Commands\DomainCommandException
+   */
+  protected function getStorage() {
+    if ($this->domain_storage) {
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: Probably more consistent to use `!is_null` here.

File: domain/src/Commands/DomainCommands.php [hunk 5]
```diff
@@ -0,0 +1,1108 @@
...
+    }
+    catch (InvalidPluginDefinitionException $e) {
+      throw new DomainCommandException('Unable to get domain: bad storage', $e);
+    }
+
+    return $this->domain_storage;
+  }
+
+  /**
+   * Lookup a string representing a domain config into that config, or throw.
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: THis sentence doesn't read very well. I don't know what it means.

File: domain/src/Commands/DomainCommands.php [hunk 6]
```diff
@@ -0,0 +1,1108 @@
...
+   * 
+   * @param string $argument
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: Does this actually handle both cases? It looks like it only handles machine_name.

File: domain/src/Commands/DomainCommands.php [hunk 7]
```diff
@@ -0,0 +1,1108 @@
...
+    if (!$domain) {
+      throw new DomainCommandException(
+        dt('Domain record could not be found from "!a".', ['!a' => $argument])
+      );
+    }
+    return $domain;
+  }
+
+  /**
+   * Filter a list of domains by excluding
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: By excluding what?

File: domain/src/Commands/DomainCommands.php [hunk 8]
```diff
@@ -0,0 +1,1108 @@
...
+   *   List of domains.
+   * @param int[] $exclude
+   *   List of domain ID numbers to exclude from the list.
+   *
+   * @return array
+   */
+  protected function filterDomains($domains, $exclude, $initial = []) {
+    $list = $initial;
+    foreach ($domains as $one) {
+      // Not the domain we're deleting, and not the default.
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: Note.

File: domain/src/Commands/DomainCommands.php [hunk 9]
```diff
@@ -0,0 +1,1108 @@
...
+   *
+   * @param \Drupal\domain\DomainInterface $domain
+   *
+   * @return string
+   */
+  protected function checkHTTPResponse(DomainInterface $domain, $validate_url) {
+    // Ensure the url is rebuilt.
+    if ($validate_url) {
+      $this->validateDomain($domain->getHostname());
+    }
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: See `DomainGetResponseTest` for the expected behavior here.

File: domain/src/Commands/DomainCommands.php [hunk 10]
```diff
@@ -0,0 +1,1108 @@
...
+      if (!$domain instanceof DomainInterface) {
+        throw new StorageException('deleting domains: value is not a domain');
+      }
+
+      $hostname = $domain->getHostname();
+      try {
+        $domain->delete();
+      }
+      catch (EntityStorageException $e) {
+        throw new DomainCommandException(dt('Unable to delete domain: @dom',
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: Let's not shorthand `@domain`

File: domain/src/Commands/DomainCommands.php [hunk 11]
```diff
@@ -0,0 +1,1108 @@
...
+    $validator = \Drupal::service('domain.validator');
+    return $validator->validate($hostname);
+  }
+
+  /**
+   * Deletes a domain record.
+   *
+   * @param DomainInterface[] $domains
+   *   The domain_id to delete. Pass 'all' to delete all records.
+   * @param string[] $delete_options
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php:186]: This is where we have to consider using interactive mapping of the fields. In D7 it was a little easier because we had fewer fields.

The idea of `$domain_options` is to shortcut the prompts. Ideally, the options / prompts would be:

* Domain Administrators: `DOMAIN_ADMIN_FIELD`
* A keyed list of fields defined by other modules

File: domain/src/Commands/DomainCommands.php [hunk 12]
```diff
@@ -0,0 +1,1108 @@
...
+    $ids = $this->enumerateDomainEntities($entity_type, $old_domain);
+
+    // @TODO is there a problem loading many, possibly big, entities in one go?
+    $entities = $entity_storage->loadMultiple($ids);
+    $count = 0;
+
+    /** @var \Drupal\Core\Entity\EntityInterface $entity */
+    foreach($entities as $entity) {
+      $changed = FALSE;
+      $value = $entity->get(DOMAIN_ACCESS_FIELD)->value();
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: Written well, this field would be an argument so this routine could run for all of our entity fields.

File: domain/src/Commands/DomainCommands.php [hunk 13]
```diff
@@ -0,0 +1,1108 @@
...
+    }
+  }
+
+  /**
+   * Reassign entities of the supplied type to the $policy domain.
+   *
+   * @return string[]
+   *   List of entities supporting domain access.
+   */
+  protected function findDomainEnabledEntities() {
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: This is a really smart abstraction 👍

File: domain/src/Commands/DomainCommands.php [hunk 14]
```diff
@@ -0,0 +1,1108 @@
...
+    foreach ($domains as $domain) {
+      $row = [];
+      foreach ($keys as $key) {
+        $v = $domain->get($key);
+        switch($key) {
+          case 'status':
+            if (($options['inactive'] && $v) || ($options['active'] && !$v)) {
+              continue 3; // switch, for, for
+            }
+            $v = ($v) ? dt('Active') : dt('Inactive');
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: I find this shorthand `$v = ($v)` hard to parse. Id rather use `!empty($v) ? dt('Active') : dt('Inactive');`

File: domain/src/Commands/DomainCommands.php [hunk 15]
```diff
@@ -0,0 +1,1108 @@
...
+   * @usage drush domain-test
+   * @usage drush domain-test example.com
+   * @usage drush domain-test 1
+   *
+   * @command domain:test
+   * @aliases domain-test
+   * 
+   * @throws \Drupal\domain\Commands\DomainCommandException
+   */
+  public function test($domain_id, array $options = ['base_path' => null]) {
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php:885]: See `checkHTTPResponse` above, which should probably contain most of this checking code.

File: domain/src/Commands/DomainCommands.php [hunk 16]
```diff
@@ -0,0 +1,1108 @@
...
+    $domain_storage = $this->getStorage();
+
+    // TODO: This won't work in a subdirectory without a parameter.
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php:888]: Not sure this is still true?

File: domain/src/Commands/DomainCommands.php [hunk 17]
```diff
@@ -0,0 +1,1108 @@
...
+   *   The machine-readable name to use for the domain.
+   * @usage drush domain-machine-name example.com foo
+   * @usage drush domain-machine-name 1 foo
+   *
+   * @command domain:machine-name
+   * @aliases domain-machine-name
+   *
+   * @throws \Drupal\domain\Commands\DomainCommandException
+   */
+  public function machineName($domain_id, $machine_name) {
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: This was a super-admin tool in D7 when we didn't use machine name for data storage, we just used numeric id. The field data now uses machine_name as the value, which makes this command very volatile. 

If we change a machine name, we have to change any field storage reference to the old machine name.

`enumerateDomainEntities` can help with that check.

File: domain/src/Commands/DomainCommands.php [hunk 18]
```diff
@@ -0,0 +1,1108 @@
...
+    foreach($field_map as $type => $fields) {
+      if (array_key_exists(DOMAIN_ACCESS_FIELD, $fields)) {
+        $entities[] = $type;
+      }
+    }
+    return $entities;
+  }
+
+  /**
+   * Enumerate entities of the supplied type and domain to the $policy domain.
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: Probably also needs a $field argument.

File: domain/src/Commands/DomainCommands.php [hunk 19]
```diff
@@ -0,0 +1,1108 @@
...
+    }
+    else {
+      $this->logger()->info(dt('Nothing to do?'));
+      return;
+    }
+    // Create list of domain (entity) ids to delete.
+    $delete_domain_ids = array_map(function($v){return $v->id;}, $domains);
+    //endregion
+
+    //region Get content disposition from configuration and validate.
```
Comment (agentrickard) [inline domain/src/Commands/DomainCommands.php]: These regions likely need to be calls to the registry we discussed, where each field acts in it's own method. Those methods like go in `DomainElementManager` and are modeled on `getFieldValues`

File: domain/src/Commands/DomainCommands.php (hunk 14 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Done

File: domain/src/Commands/DomainCommands.php (hunk 13 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: cool

File: domain/src/Commands/DomainCommands.php (hunk 12 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Refactored... was in some of my newest code.

File: domain/src/Commands/DomainCommands.php (hunk 11 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php:186]: Going to have to think further about this area.

File: domain/src/Commands/DomainCommands.php (hunk 10 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Done

File: domain/src/Commands/DomainCommands.php (hunk 7 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Sorted

File: domain/src/Commands/DomainCommands.php (hunk 6 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: As discussed, yes. Improved sentence.

File: domain/src/Commands/DomainCommands.php (hunk 5 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Sorted.

File: domain/src/Commands/DomainCommands.php (hunk 4 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Done.

File: domain/src/Commands/DomainCommands.php (hunk 16 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php:888]: Haven't done anything of note for this fn, just the basics. Need to look into it further.

File: domain/src/Commands/DomainCommands.php (hunk 17 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: ok, might park that one and disable this fn. Sounds like a reassign all over again...

File: domain/src/Commands/DomainCommands.php (hunk 19 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: I'll have a think about this too.

File: domain/src/Commands/DomainCommands.php (hunk 18 above)
Comment (rivimey) [reply domain/src/Commands/DomainCommands.php]: Done.

**Comment by agentrickard** [comment]
I think I fixed the intermittent ResponseTest failures with commit 

https://github.com/agentrickard/domain/pull/434/commits/15275090e133364906f9e27d6d27e0721753402e

So you might want to cherry-pick that over.
